File: src/lib/wow/db/caches/init-cache.ts
/**
 * Model cache initialization, ported from wow.export (src/js/db/caches/init-cache.js).
 */
import { doOnce } from '../../formats/generics';
import { initializeCreatureData, resetCreatureCache } from './db-creatures';
import { initializeItemDisplays, resetItemDisplayCache } from './db-item-displays';
import { initializeModelFileData, resetModelFileDataCache } from './db-model-file-data';
import { initializeTextureFileData, resetTextureFileDataCache } from './db-texture-file-data';

/** Load DB2 skin caches in dependency order (lower peak memory than parallel init). */
export const ensureModelCachesInitialized = doOnce('ensureModelCachesInitialized', async () => {
  await initializeModelFileData();
  await initializeTextureFileData();
  await Promise.all([
    initializeItemDisplays(),
    initializeCreatureData(),
  ]);
});

export function initModelCaches(): Promise<unknown>[] {
  return [ensureModelCachesInitialized()];
}

export function resetDbCaches(): void {
  resetCreatureCache();
  resetItemDisplayCache();
  resetModelFileDataCache();
  resetTextureFileDataCache();
}

export default { initModelCaches, ensureModelCachesInitialized, resetDbCaches };

File: golang/internal/wow/db/caches/init_cache.go
package caches

import (
	"context"
	"sync"
)

var (
	modelInitOnce     sync.Once
	modelInitErr      error
	modelCachesLoaded bool
)

// EnsureModelCachesInitialized loads DB2 skin caches in dependency order.
func EnsureModelCachesInitialized(ctx context.Context) error {
	modelInitOnce.Do(func() {
		if err := InitializeModelFileData(ctx); err != nil {
			modelInitErr = err
			return
		}
		if err := InitializeTextureFileData(ctx); err != nil {
			modelInitErr = err
			return
		}
		if err := InitializeItemDisplays(ctx); err != nil {
			modelInitErr = err
			return
		}
		if err := InitializeCreatureData(ctx); err != nil {
			modelInitErr = err
			return
		}
		modelCachesLoaded = true
	})
	return modelInitErr
}

// ResetDBCaches clears all DB caches.
func ResetDBCaches() {
	ResetCreatureCache()
	ResetItemDisplayCache()
	ResetModelFileDataCache()
	ResetTextureFileDataCache()
	modelInitOnce = sync.Once{}
	modelInitErr = nil
	modelCachesLoaded = false
}

// ModelCachesInitialized reports whether model DB caches loaded successfully.
func ModelCachesInitialized() bool {
	return modelCachesLoaded
}
